File: Cpp/0491_non_decreasing_subsequences.cpp
// https://leetcode.cn/problems/non-decreasing-subsequences/

class Solution {
public:
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        nums_ = nums;
        unordered_set<int> check;
        for (int i = 0; i < nums.size(); ++i) {
            int num = nums[i];
            if (check.find(num) == check.end()) {
                check.insert(num);
                vector<int> cur = {num};
                dfs(i+1, cur);
            }
        }
        return res_;
    }

    void dfs(int idx, vector<int> &cur) {
        unordered_set<int> check;
        for (int i = idx; i < nums_.size(); ++i) {
            int num = nums_[i];
            if (check.find(num) == check.end() && num >= cur.back()) {
                check.insert(num);
                cur.push_back(num);
                res_.push_back(cur);
                dfs(i+1, cur);
                cur.pop_back();
            }
        }
    }

    vector<int> nums_;
    vector<vector<int>> res_;
};

File: Python3/0491_non_decreasing_subsequences.py
# https://leetcode.cn/problems/non-decreasing-subsequences/

class Solution:
    def findSubsequences(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: list[list[int]]
        """
        self.res = []
        self.nums = nums
        s = set()
        for i, num in enumerate(nums):
            if num not in s:
                s.add(num)
                self.dfs(i+1, [num])
        return self.res
    
    def dfs(self, idx, tmp):
        s = set()
        for i in range(idx, len(self.nums)):
            num = self.nums[i]
            if num in s or num < tmp[-1]:
                continue
            
            s.add(num)
            cur = tmp + [num]
            self.res.append(cur)
            self.dfs(i+1, cur)